Textfiled support cursor v1.2

perminovVS opened the pull request:
Add StringUtils::StringUTF8 for easy manipulate utf8 string.
If TextField active(focused input text) show text cursor( default '|',
change TextFieldTTF::setCursorChar ).
Show/Hide interval CURSOR_TIME_SHOW_HIDE
Change position text cursor control key: KEY_HOME, KEY_END,
KEY_LEFT_ARROW, KEY_RIGHT_ARROW
CC_PLATFORM_MAC, CC_PLATFORM_WIN32, CC_PLATFORM_LINUX default use
cursor( For change state TextFieldTTF::setCursorUse )
Desktop implement dispatchControlKey
WinRT support key repeat and implement dispatchControlKey
Text cursor worked only on LabelType::TTF

new version it: #13903 , for easy merge

Example use: http://youtu.be/g6jpmaip1fY

perminovVS committed:
StringUTF8 - Utf8 sequence

Store each utf8 chars as std::string, for easily manipulated

perminovVS committed:
Label '\b' - Next char not change x position

Add TextFormatter for easy use '\n', '\r', '\b'

perminovVS committed:
TextFieldTTF support cursor

If TextField active(focused input text) show text cursor( default '|',
change TextFieldTTF::setCursorChar ).
Show/Hide interval CURSOR_TIME_SHOW_HIDE
Change position text cursor control key: KEY_HOME, KEY_END,
KEY_LEFT_ARROW, KEY_RIGHT_ARROW
CC_PLATFORM_MAC,  CC_PLATFORM_WIN32, CC_PLATFORM_LINUX default use
cursor( For change state TextFieldTTF::setCursorUse )
Text cursor worked only on LabelType::TTF

perminovVS committed:
Desktop implement dispatchControlKey

And fix bag control char for OSX onGLFWCharCallback
glow send on CharCallback character control key unicode.

perminovVS committed:
WinRT support key repeat and implement dispatchControlKey

zilongshanren commented:
@perminovVS  :+1:  Sweet!

zilongshanren commented:
@perminovVS 
It seems the cursor support is only available for desktop?

zilongshanren commented on cocos/base/ccUTF8.h:
```diff
@@ -142,6 +142,43 @@ CC_DLL unsigned int getIndexOfLastNotChar16(const std::vector<char16_t>& str, ch
+    ~StringUTF8();
+
+    std::size_t length() const;
+    void set(const std::string& newStr);
```
The `set` API seems not intuitive.. What's the intent of this API?

perminovVS replied:
@zilongshanren set new value from std::string

zilongshanren replied:
@perminovVS  It's `append` or `replace`?

perminovVS replied:
@zilongshanren set new values, as replace all

zilongshanren replied:
so @perminovVS  how about about renaming it to `replace` or sth  else?

zilongshanren commented on cocos/base/ccUTF8.h:
```diff
@@ -142,6 +142,43 @@ CC_DLL unsigned int getIndexOfLastNotChar16(const std::vector<char16_t>& str, ch
+
+    StringUTF8();
+    StringUTF8(const std::string& newStr);
+    explicit StringUTF8(const StringUTF8& copyStrUtf8);
```
if we define a user defined copy constructor, we might need to use the "Rule of Three" or "Rule of Five"...

zilongshanren commented on cocos/base/ccUTF8.h:
```diff
@@ -142,6 +142,43 @@ CC_DLL unsigned int getIndexOfLastNotChar16(const std::vector<char16_t>& str, ch
+* Store all utf8 chars as std::string
+* Build from std::string
+*/
+class CC_DLL StringUTF8
```
One more question: Do we really need the StringUTF8 encapsulation? I think we have already provided many utf8, utf16 string conversion functions, why bother to provide another StringUTF8 class?
Thanks.

perminovVS replied:
StringUTF8 for easy manipulate sequence utf8 chars;
example from TextFieldTTF

```
StringUtils::StringUTF8 stringUTF8;

stringUTF8.set(_inputText);
stringUTF8.insert(_cursorPosition, insert);
```

perminovVS commented:
@zilongshanren CC_PLATFORM_MAC, CC_PLATFORM_WIN32, CC_PLATFORM_LINUX default use
cursor, other platform need change state TextFieldTTF::setCursorUse(true)

zilongshanren commented:
@perminovVS 
Thanks for the clarification, I will do more test to this PR.

zilongshanren commented on cocos/2d/CCTextFieldTTF.h:
```diff
@@ -230,6 +236,32 @@ class CC_DLL TextFieldTTF : public Label, public IMEDelegate
+    * Set enable cursor use.
+    * @js NA
+    */
+    void setCursorUse(bool value);
```
`setCursorEnabled` might be better than `setCursorUse`.

zilongshanren commented on cocos/2d/CCTextFieldTTF.h:
```diff
@@ -252,6 +287,21 @@ class CC_DLL TextFieldTTF : public Label, public IMEDelegate
     bool _secureTextEntry;
 
+    // Need use cursor
+    bool _cursorUse;
```
change `_cursorUse` to `_cursorEnabled`.

zilongshanren commented:
@perminovVS 
Now I can't click to change cursor position, only arrow key is working. Thoughts?

perminovVS commented:
@zilongshanren yes, need manualy call setCursorFromPoint.
Need listen mouse or touch event and call it.

In UITextField can add this functional.

perminovVS committed:
Remarks zilongshanren

Copy constructor remove from StringUTF8
StringUTF8::set change to StringUTF8::replace
TextFieldTTF::setCursorUse change to TextFieldTTF::setCursorEnabled
TextFieldTTF::_cursorUsechange to TextFieldTTF::_cursorEnabled

perminovVS commented:
@zilongshanren review now

zilongshanren commented:
@perminovVS  I'm OK with this PR now.  Thanks.

0xJimmyBeh commented:
@perminovVS  @zilongshanren 

May i know how to use this cursor feature on UITextField?

I tried to implement with Cocos2d-JS v3.8.1 and generated the bindings.
But i can't see the cursor after `setCursorEnabled(true);` in iOS.

Thank you.

siarsky commented:
@Zinitter 
I've looked into the cocos2d_tests project - the class TextInputTests is a good starting point with one very crucial error regarding a new cursor functionality.

The TextInputTests contains a method onEnter() - which calls
`KeyboardNotificationLayer::onEnter();`
but does not call:
`pTextField->onEnter()`
in which the schedule "update(float delta)" method will be resumed (default = paused). If onEnter() is not called on TextFieldTTF, the TextFieldTTF::update(float delta) method will not be called and the cursor will not blink.

Maybe you just do not see the cursor because it is a non-blinkin pipe sign (|) at the end of the string due to not calling TextFieldTTF::onEnter()...

After calling onEnter the cursor functionality works for me in iOs and Android as well.
